libfeed/feed-model-ordering.c

```diff
@@ -0,0 +1,401 @@
+    }
+
+  return convert_nested_hashtables_to_indexable_descriptors_map (map);
+}
+
+#define INNER_OUTER_LENGTH 2
+#define INNER_TYPES_LENGTH 4
+
+static EosDiscoveryFeedCardStoreType card_type_indices_ordering[INNER_OUTER_LENGTH][INNER_TYPES_LENGTH] = {
+  {
+    EOS_DISCOVERY_FEED_CARD_STORE_TYPE_NEWS_CARD,
+    EOS_DISCOVERY_FEED_CARD_STORE_TYPE_ARTICLE_CARD,
+    EOS_DISCOVERY_FEED_CARD_STORE_TYPE_VIDEO_CARD,
+    EOS_DISCOVERY_FEED_CARD_STORE_TYPE_ARTWORK_CARD,
+  },
+  {
+    EOS_DISCOVERY_FEED_CARD_STORE_TYPE_NEWS_CARD,
+    EOS_DISCOVERY_FEED_CARD_STORE_TYPE_VIDEO_CARD,
+    EOS_DISCOVERY_FEED_CARD_STORE_TYPE_ARTICLE_CARD,
+    EOS_DISCOVERY_FEED_CARD_STORE_TYPE_ARTWORK_CARD
+  }
+};
+
+static inline guint
+lookup_card_limit (EosDiscoveryFeedCardStoreType type)
+{
+  switch (type)
+    {
+      case EOS_DISCOVERY_FEED_CARD_STORE_TYPE_NEWS_CARD:
+        return 5;
+      case EOS_DISCOVERY_FEED_CARD_STORE_TYPE_ARTICLE_CARD:
+      case EOS_DISCOVERY_FEED_CARD_STORE_TYPE_VIDEO_CARD:
+      case EOS_DISCOVERY_FEED_CARD_STORE_TYPE_ARTWORK_CARD:
+        return 1;
+      default:
+        break;
+    }
+
+  /* Default case */
+  return 0;
+}
+
+static GHashTable *
+make_app_indices_for_card_type (const EosDiscoveryFeedCardStoreType *types)
+{
+  g_autoptr(GHashTable) app_indices_for_card_types = g_hash_table_new_full (g_direct_hash,
+                                                                            g_direct_equal,
+                                                                            NULL,
+                                                                            NULL);
+  const EosDiscoveryFeedCardStoreType *iter = types;
+
+  for (; *iter != EOS_DISCOVERY_FEED_CARD_STORE_TYPE_UNSET; ++iter)
+    g_hash_table_insert (app_indices_for_card_types, GINT_TO_POINTER (*iter), 0);
+
+  return g_steal_pointer (&app_indices_for_card_types);
+}
+
+#define CARDS_LIMIT 14
+
+/**
+ * eos_discovery_feed_arrange_orderable_models:
+ * @unordered_orderable_models: (element-type EosDiscoveryFeedOrderableModel): The
+ *                              models to to order. The output list may be truncated
+ *                              as a result of the reordering operation.
+ *
+ * Reorder the models in the order that they should be displayed in the feed.
+ * The pattern here is not predictable by the user but shuffles through a range
+ * of content adequately and is not in fact random, but changes day by day.
+ *
+ * The general approach here is to insert a news card, article card, video card
+ * and artwork card from a diferent app, with the word-quote card being inserted
+ * second. Then a news card, video card, article card and artwork card will
+ * be inserted and the pattern will repeat itself. There is a limit of one card
+ * per app, except for news cards where there is a limit of 5 cards.
+ *
+ * Returns: (transfer container) (element-type EosDiscoveryFeedOrderableModel): The
```

Comment:
So it seems that in order to make GPtrArray work with language bindings you need to use `transfer container` as opposed to `transfer full` (which crashes). To be honest, I should really file an upstream bug about that, since it is a problem in both pygi and gjs. `transfer container` thankfully does not leak, either.

I'll adjust the other functions that return containers too in the other commits.

Reply:
Done

Reply:
(But still need to file a bug, its on my list!)